lib/Mailer/Mailjet.php

```diff
@@ -0,0 +1,210 @@
+     */
+    public function setServerToken($serverToken)
+    {
+        if (false === strpos($serverToken, ':')) {
+            throw new \InvalidArgumentException('Mailjet uses a "publicApiKey:privateApiKey" based ServerToken');
+        }
+
+        parent::setServerToken($serverToken);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getHeaders()
+    {
+        return [
+            'Accept' => 'application/json',
+            'Authorization' => sprintf('Basic %s', base64_encode($this->getServerToken())),
+            'Content-Type' => 'application/json',
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function format(MessageInterface $message)
+    {
+        $headers = [];
+        foreach ($message->getHeaders() as $name => $value) {
+            $headers[$name] = $value;
+        }
+
+        $parameters = [
+            'From' => $this->buildIdentityString($message->getFrom()),
+            'To' => $this->buildIdentityString($message->getTo()),
+            'Subject' => $message->getSubject(),
+            'Headers' => $headers,
+            'HTMLPart' => $message->getHtml(),
+            'TextPart' => $message->getText(),
+            'ReplyTo' => $this->buildIdentityString($message->getReplyTo()),
+        ];
+
+        $attachments = $this->getFiles($message);
+
+        if ($attachments['attached']) {
+            $parameters['Attachments'] = $attachments['attached'];
+        }
+
+        if ($attachments['inlined']) {
+            $parameters['InlinedAttachments'] = $attachments['inlined'];
+        }
+
+        if ($message instanceof TaggableInterface) {
+            $tag = $message->getTag();
+
+            if (is_array($tag)) {
+                $tag = reset($tag);
+            }
+
+            $parameters['EventPayload'] = $tag;
+        }
+
+        return json_encode(['Messages' => [array_filter($parameters)]]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function handle(ResponseInterface $response)
+    {
+        $statusCode = $response->getStatusCode();
+        $httpException = new HttpException($statusCode, $response->getStatusText());
+
+        if (!in_array($statusCode, [400, 401, 403], true)) {
+            throw $httpException;
+        }
+
+        $error = json_decode($response->getContent(), true);
+
+        if (isset($error['ErrorMessage'])) {
+            throw new ApiException($error['ErrorMessage'], $httpException);
+        }
+
+        $errorMessages = [];
+        foreach ($error['Messages'] as $message) {
+            if ('error' !== $message['Status']) {
+                continue;
+            }
+
+            foreach ($message['Errors'] as $mailError) {
+                $errorMessages[] = $mailError['ErrorMessage'];
+            }
+        }
+
+        throw new ApiException(implode(', ', $errorMessages), $httpException);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function buildIdentityString($identities)
```

Comment:
the name is really weird, as this can return arrays most of the time

Reply:
I'm just overriding the base class' function for consistency. Maybe it should be renamed at this level ?

Reply:
well, overriding the parent method and changing the return type to something else than a string is even violating the OOP rules.

Just use a different name reflecting what it actually does (and make it private btw)